File: app/src/main/java/cm/aptoide/pt/autoupdate/AutoUpdateService.java
package cm.aptoide.pt.autoupdate;

import cm.aptoide.pt.dataprovider.exception.NoNetworkConnectionException;
import retrofit2.http.GET;
import retrofit2.http.Path;
import rx.Observable;
import rx.Single;

public class AutoUpdateService {

  private final Service service;
  private final String packageName;
  private final String autoUpdateStoreName;
  private boolean loading;

  public AutoUpdateService(Service service, String packageName, String autoUpdateStoreName) {
    this.service = service;
    this.packageName = packageName;
    this.autoUpdateStoreName = autoUpdateStoreName;
  }

  public Single<AutoUpdateModel> loadAutoUpdateModel() {
    if (loading) {
      return Single.just(new AutoUpdateModel(true));
    }
    return service.getJsonResponse(autoUpdateStoreName)
        .doOnSubscribe(() -> loading = true)
        .doOnUnsubscribe(() -> loading = false)
        .doOnTerminate(() -> loading = false)
        .flatMap(jsonResponse -> Observable.just(
            new AutoUpdateModel(jsonResponse.getVersioncode(), jsonResponse.getUri(),
                jsonResponse.getMd5(), jsonResponse.getMinSdk(), packageName, false)))
        .onErrorReturn(throwable -> createErrorAutoUpdateModel(throwable))
        .toSingle();
  }

  private AutoUpdateModel createErrorAutoUpdateModel(Throwable throwable) {
    if (throwable instanceof NoNetworkConnectionException) {
      return new AutoUpdateModel(AutoUpdateModel.Error.NETWORK);
    } else {
      return new AutoUpdateModel(AutoUpdateModel.Error.GENERIC);
    }
  }

  public interface Service {
    @GET("latest_version_{storeName}.json") Observable<AutoUpdateJsonResponse> getJsonResponse(
        @Path(value = "storeName") String storeName);
  }
}

File: app/src/main/java/cm/aptoide/pt/autoupdate/AutoUpdateService.kt
package cm.aptoide.pt.autoupdate

import cm.aptoide.pt.dataprovider.exception.NoNetworkConnectionException
import retrofit2.http.GET
import retrofit2.http.Path
import rx.Observable
import rx.Single

class AutoUpdateService(private val service: Service, private val packageName: String,
                        private val clientSdkVersion: Int) {

  private var loading = false

  fun loadAutoUpdateModel(): Single<AutoUpdateModel> {
    if (loading) {
      return Single.just(AutoUpdateModel(loading = true))
    }
    return service.getAutoUpdateResponse(packageName, clientSdkVersion)
        .doOnSubscribe { loading = true }
        .doOnUnsubscribe { loading = false }
        .doOnTerminate { loading = false }
        .flatMap {
          Observable.just(AutoUpdateModel(it.versioncode, it.uri, it.md5, it.minSdk, packageName))
        }
        .onErrorReturn { createErrorAutoUpdateModel(it) }
        .toSingle()

  }

  private fun createErrorAutoUpdateModel(throwable: Throwable?): AutoUpdateModel? {
    return when (throwable) {
      is NoNetworkConnectionException -> AutoUpdateModel(status = Status.ERROR_NETWORK)
      else -> AutoUpdateModel(status = Status.ERROR_GENERIC)
    }
  }
}

interface Service {
  @GET("apks/package/autoupdate/get/package_name={package_name}/sdk={client_sdk_version}")
  fun getAutoUpdateResponse(
      @Path(value = "package_name") packageName: String, @Path(value = "client_sdk_version")
      clientSdkVersion: Int): Observable<AutoUpdateJsonResponse>
}